### ISSUE TITLE
Extreme Memory Consumption and connect to Webserver times out

### ISSUE THREAD

**Issue by DirkFries** [cblomart/vsphere-graphite#101]
Hi,

I am using vsphere-graphite in combination with Prometheus. During the last weeks I could observe 2 strange behaviours that might cohere.

My vsphere-graphite Version is:
Version information: NA - aba709b@HEAD (dirty)

1.  The longer vsphere graphite the more memory does it consume. The maximum that I have seen was 3,5 GB resident memory

2. From time to time the Webserver of vsphere-graphite that holds the metrics does not respond. Very strange : In such a case neither a restart of vsphere-graphite nor a reboot does help. After same time the Webserver works again an Prometheus can collect the metrics.

My Metrics config looks like that:

 "Metrics": [
    {
      "ObjectType": [ "VirtualMachine", "HostSystem" ],
      "Definition": [
        { "Metric": "cpu.usage.average", "Instances": "" },
        { "Metric": "cpu.usage.maximum", "Instances": "" },
        { "Metric": "cpu.usagemhz.average", "Instances": "" },
        { "Metric": "cpu.usagemhz.maximum", "Instances": "" },
        { "Metric": "cpu.totalCapacity.average", "Instances": "" },
        { "Metric": "mem.usage.average", "Instances": "" },
        { "Metric": "mem.usage.maximum", "Instances": "" },
        { "Metric": "mem.consumed.average", "Instances": "" },
        { "Metric": "mem.consumed.maximum", "Instances": "" },
        { "Metric": "mem.active.average", "Instances": "" },
        { "Metric": "mem.active.maximum", "Instances": "" },
        { "Metric": "mem.vmmemctl.average", "Instances": "" },
        { "Metric": "mem.vmmemctl.maximum", "Instances": "" },
        { "Metric": "mem.totalCapacity.average", "Instances": "" }
      ]
    },
    {
      "ObjectType": [ "VirtualMachine" ],
      "Definition": [
        { "Metric": "virtualDisk.totalWriteLatency.average", "Instances": "*" },
        { "Metric": "virtualDisk.totalReadLatency.average", "Instances": "*" },
        { "Metric": "virtualDisk.numberReadAveraged.average", "Instances": "*" },
        { "Metric": "virtualDisk.numberWriteAveraged.average", "Instances": "*" },
        { "Metric": "cpu.ready.summation", "Instance": ""}
      ]
    },
    {
      "ObjectType": [ "HostSystem" ],
      "Definition": [
        { "Metric": "disk.maxTotalLatency.latest", "Instances": "" },
        { "Metric": "disk.numberReadAveraged.average", "Instances": "*" },
        { "Metric": "disk.numberWriteAveraged.average", "Instances": "*" },
        { "Metric": "cpu.usage.average", "Instances": "*" },
        { "Metric": "cpu.utilization.average", "Instances": "*" },
        { "Metric": "mem.usage.average", "Instances": "*" }
      ]

All in all there are about 1800 VMs on about 80 Hosts.

Bye, Dirk

**Comment by cblomart** [comment]
Hi Dirk,

I know this might not change much but could you run from https://github.com/cblomart/vsphere-graphite/releases/download/v0.8a/vsphere-graphite_linux_amd64.tgz

Would you also have the logs from vsphere-graphite when you are in those situations ? I would like to check if it reports the correct amounts.

Cedric

**Comment by DirkFries** [comment]
Hi Cedric,

I just download the version you have mentioned and I will use this one in the future. 
The error appears only after some time. So it might last some weeks until I can report.

As soon as I have new information I will report.

Thanks a lot !

**Comment by cblomart** [comment]
I checked vsphere-graphite usage @work and seen some similar behaviour... mainly the /metrics endpoint not responding. Additionally i've seen some errors from prometheus about extra headers written twice or so... 
I restarted vsphere graphite with thinprometheus that doesn't depend on the prometheus general client.
I guess i'll monitor that to verify if such behavior subsits.

Additionally i discussed with the prometheus admin and realised he was polling vsphere graphite every 5 seconds... please don't do that... basicaly vsphere graphite will pull the last minut fro vsphere (real time data stored in memory in esxi). 
**Best practice should be to poll every minut**

**Comment by cblomart** [comment]
So while running vsphere-graphite for a few days i notice that then number of go routines is increasing.
In my test the number increased from 7 routines to 18 routrines in 5 days.

A go routine is lightweight and shouldn't cause too much issues, still accumulating them is not a good idea.

I'll set some code to have some traces (this may get verbose in the logs).

**Comment by DirkFries** [comment]
Hi cblomart,

good to see that I am not the only one getting this behavior.
Since my last restart of vsphere-graphite about 11 days ago the resident memory as risen from about 128 MB to ~680MB. 

In my case Prometheus is polling every minute. I have chosen this intervall because vsphere-graphite needs about 19 seconds to get all the metrics from my vspheres.

Verbose logs wouldn't be a problem at the moment, because my installation is not in 100%production mode at the moment.

**Comment by cblomart** [comment]
I don’t see such increase in our logs for memory, just the go routines increasing.

Are you using the “prometheus” part or “thinprometheus” backend? 

As indicated I switched ours to the latter as i saw some strange behavior not directly linked to memory usage.

I will still continue investigating goroutines.

**Comment by cblomart** [comment]
short update... 

I am running a modified version with more info on go routines (see goroutines branch; docker image cblomart/vsphere-graphite:3c8681a)

For the last day no increase whatsoever (go routines and memory).

**Comment by DirkFries** [comment]
In the meantime my running process has reached 800 MB resident memory, but webserver still works. 

Do you think I should switch to the version you have mentioned or do you think I should wait until webserver times out to get (maybe) some errors in the logfile ?

**Comment by cblomart** [comment]
Do you run the version with extensive logging of go routines?

I would like to first confirm that we have a valid workaround. Could you update your configuration to use the thinprometheus driver. 
Changing the name is all you need to do.

With this we can check:
* if memory consumption grows
* if the metric endpoint becomes unresponsive after a while

**Comment by cblomart** [comment]
Looking at thinprometheus i think there might be some incomplete handling of timeouts.
https://github.com/cblomart/vsphere-graphite/blob/aba709b743ab241bd2e4f123847c8813be25a3a2/backend/thinprometheusclient.go#L96

I need to check how this is done for the standard prometheus

**Comment by DirkFries** [comment]
Just switched to thinprometheus.

Right before the switch memory consumotion was at 1,125 GB and 72 go routines.
After then switch memory consumption is at 200 MB and 15 go routines.

I'll keep you updated how thinprometheus behaves.

**Comment by cblomart** [comment]
Ok i reveiwed the timeout management of the prometheus implementations in vsphere-graphite:b7c26b7 (goroutines branch).

I did see some concurency issues (end of cycle not happening) but wasn't able to reproduce after having increased the timeouts.

I am montoring the situation from now on.

@DirkFries : what is the situation on your side? If there is still an increase of go routines, could you test with the mentionned version?

Cédric

**Comment by DirkFries** [comment]
Hi Cédric,

I had to restart the process 4 days ago because I had to add some metrics. 

At the moment it is consuming 285 MB resident RAM. I think we will have to wait for about 2 weeks at least to see any difference.

**Comment by DirkFries** [comment]
Sorry, forgot to mention.

Go routines in logfile are stable at 19

I will try the new version either today or tomorrow

**Comment by adelbot** [comment]
It seem that we have the same problem with memory and prometheus.

Logs before restart
Oct 21 18:46:22 weya0332 vsphere-graphite-prometheus: 2019/10/21 18:46:22 memory usage: sys=946.7M alloc=326M
Oct 21 18:46:22 weya0332 vsphere-graphite-prometheus: 2019/10/21 18:46:22 go routines: 53

logs after restart 
Oct 21 18:50:48 weya0332 vsphere-graphite-prometheus: 2019/10/21 18:50:48 memory usage: sys=135.8M alloc=22.2M
Oct 21 18:50:48 weya0332 vsphere-graphite-prometheus: 2019/10/21 18:50:48 go routines: 9

My config :
{
  "Domain": ".contoso.com",
  "Interval": 60,
  "VCenters": [
    { "Username": "xxxx", "Password": "xxxxxxx", "Hostname": "xxxxxxxxxxxxxxxx" }
  ],
  "Backend": {
    "Type": "Thinprometheus",
    "Hostname": "127.0.0.1",
    "Port":     5667
  },
  "Metrics": [
    {
      "ObjectType": [ "VirtualMachine" ],
      "Definition": [
        { "Metric": "cpu.usage.none", "Instance": "*"},
        { "Metric": "cpu.usage.maximum", "Instance": "*"},
        { "Metric": "cpu.usage.average", "Instance": "*"},
        { "Metric": "cpu.used.summation", "Instance": ""},
        { "Metric": "cpu.idle.summation", "Instance": ""},
        { "Metric": "mem.active.maximum", "Instances": "" },
        { "Metric": "virtualDisk.read.average", "Instances": "" },
        { "Metric": "virtualDisk.write.average", "Instances": "" },
        { "Metric": "virtualDisk.numberReadAveraged.average", "Instances": "*" },
        { "Metric": "virtualDisk.numberWriteAveraged.average", "Instances": "*" },
        { "Metric": "virtualDisk.totalWriteLatency.average", "Instances": "" },
        { "Metric": "net.bytesTx.average", "Instance": ""},
        { "Metric": "net.bytesRx.average", "Instance": ""},
        { "Metric": "power.energy.summation", "Instance": ""}
      ]
    }
  ]
}

**Comment by adelbot** [comment]
Hi All,

After a few hours 

Oct 22 11:47:39 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:47:39 go routines: 14
Oct 22 11:47:40 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:47:40 thinprom: recieve delay 1865
Oct 22 11:47:40 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:47:40 thinprom: last recieved delay 2448
Oct 22 11:47:40 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:47:40 thinprom: recieve done
Oct 22 11:48:39 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:48:39 memory usage: sys=202.8M alloc=22.2M
Oct 22 11:48:39 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:48:39 go routines: 14
Oct 22 11:48:40 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:48:40 thinprom: recieve delay 1850
Oct 22 11:48:40 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:48:40 thinprom: last recieved delay 2431
Oct 22 11:48:40 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:48:40 thinprom: recieve done
Oct 22 11:50:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:50:07 thinprom: recieve delay 0
Oct 22 11:50:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:50:07 thinprom: last recieved delay 0
Oct 22 11:50:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:50:07 thinprom: recieve incomplete
Oct 22 11:51:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:51:07 thinprom: recieve delay 0
Oct 22 11:51:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:51:07 thinprom: last recieved delay 0
Oct 22 11:51:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:51:07 thinprom: recieve incomplete
Oct 22 11:52:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:52:07 thinprom: recieve delay 0
Oct 22 11:52:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:52:07 thinprom: last recieved delay 0
Oct 22 11:52:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:52:07 thinprom: recieve incomplete
Oct 22 11:53:46 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:53:46 thinprom: recieve delay 0
Oct 22 11:53:46 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:53:46 thinprom: last recieved delay 0
Oct 22 11:53:46 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:53:46 thinprom: recieve incomplete
Oct 22 11:55:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:55:07 thinprom: recieve delay 0
Oct 22 11:55:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:55:07 thinprom: last recieved delay 0
Oct 22 11:55:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:55:07 thinprom: recieve incomplete
Oct 22 11:56:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:56:07 thinprom: recieve delay 0
Oct 22 11:56:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:56:07 thinprom: last recieved delay 0
Oct 22 11:56:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:56:07 thinprom: recieve incomplete
Oct 22 11:57:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:57:07 thinprom: recieve delay 2212
Oct 22 11:57:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:57:07 thinprom: last recieved delay 3039
Oct 22 11:57:07 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:57:07 thinprom: recieve incomplete
Oct 22 11:57:40 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:57:40 thinprom: recieve delay 2039
Oct 22 11:57:40 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:57:40 thinprom: last recieved delay 2880
Oct 22 11:57:40 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:57:40 thinprom: recieve incomplete
Oct 22 11:58:40 weya0332 vsphere-graphite-prometheus: 2019/10/22 11:58:40 thinprom: recieve delay 2221

I'm available for make another tests.

Best Regards.

**Comment by adelbot** [comment]
Hello,

The memory upgrade seems to come after we encounter connection problems with the vsphere.
in any case, it was obvious during a network outage

BR

**Comment by DirkFries** [comment]
Hi cblomart,

just checked again the memory usage of the process (2019/12/05 09:50:42 Version information: NA - aba709b@HEAD (dirty))
Process is running for 2 weeks and the number of go routines has risen to 247 resulting in about 8 GB Memory Usage.

During the last days I had issues with one vcenter that couldn't reach all of its hosts because of packet loss. So I guess the massive memory consumption might be a result of those problems ?

Bye Dirk

**Comment by cblomart** [comment]
@dupondje found a race condition in the handling of signals... the branch "prometheus_race" has been updated with corrective code... if you guys are still with us, 2 thinks:
* thanks for bearing with this 
* can you test the prometheus_race branch (should also have a container image ready)

**Comment by cblomart** [comment]
@DirkFries would you have time to look at this and test the version in the prometheus_race branch?

**Comment by DirkFries** [comment]
Hi Cedric,

I just tried exchanging a new built binary with the old one withour changing the config and I got this:

```
2020-03-26T13:27:33.785649+01:00 lin372 vsphere-graphite[545]: 2020/03/26 13:27:33 vcenter vcenter2.domain.local: disconnecting
2020-03-26T13:27:33.789789+01:00 lin372 vsphere-graphite[545]: panic: runtime error: makeslice: len out of range
2020-03-26T13:27:33.790114+01:00 lin372 vsphere-graphite[545]: goroutine 115 [running]:
2020-03-26T13:27:33.790355+01:00 lin372 vsphere-graphite[545]: github.com/cblomart/vsphere-graphite/vsphere.(*VCenter).Query(0xc0004a9ef8, 0x3c, 0xc000218d40, 0xd, 0x0, 0xc000212100, 0xb, 0x10, 0x0, 0x0, ...)
2020-03-26T13:27:33.790509+01:00 lin372 vsphere-graphite[545]: #011/nethome/vd/go/src/github.com/cblomart/vsphere-graphite/vsphere/vsphere.go:487 +0x3150
2020-03-26T13:27:33.790659+01:00 lin372 vsphere-graphite[545]: main.queryVCenter(0xc00021a6c0, 0x15, 0xc000218d90, 0x10, 0xc000218da0, 0x7, 0xc0002de610, 0x2, 0x2, 0xc000206c40, ...)
2020-03-26T13:27:33.790813+01:00 lin372 vsphere-graphite[545]: #011/tmp/vsphere-graphite/vsphere-graphite.go:48 +0xbf
2020-03-26T13:27:33.790959+01:00 lin372 vsphere-graphite[545]: created by main.(*Service).Manage.func1
2020-03-26T13:27:33.791104+01:00 lin372 vsphere-graphite[545]: #011/tmp/vsphere-graphite/vsphere-graphite.go:319 +0x1c3
2020-03-26T13:27:33.791941+01:00 lin372 systemd[1]: vsphere-graphite.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
2020-03-26T13:27:33.792677+01:00 lin372 systemd[1]: vsphere-graphite.service: Unit entered failed state.
2020-03-26T13:27:33.792952+01:00 lin372 systemd[1]: vsphere-graphite.service: Failed with result 'exit-code'.
2020-03-26T13:27:33.921792+01:00 lin372 systemd[1]: vsphere-graphite.service: Service RestartSec=100ms expired, scheduling restart.
2020-03-26T13:27:33.922154+01:00 lin372 systemd[1]: Stopped Vsphere Graphite Server.
2020-03-26T13:27:33.923252+01:00 lin372 systemd[1]: Started Vsphere Graphite Server.
```

I'm not sure if a colleage and me did something wrong compiling?

Bye, Dirk

**Comment by cblomart** [comment]
Thanks a lot for the test... can you include the "headers" of the logs so i can check which commit was used?

@dupondje made some change and you need to include some extra parameters in the config file....
We need to document this also (maybe on the race branch for now).

```json
  "VCenterResultLimit": 500000,
  "VCenterInstanceRatio": 3.0,
```

**Comment by DirkFries** [comment]
Hi Cedric,

here are the headers:

```
d7b919a (HEAD -> prometheus_race, origin/prometheus_race) handle case in metric defs

01d9fc8 reset timer after done

01dedd8 (origin/master, origin/HEAD, master) Merge pull request #113 from cblomart/build

b046746 correct S3 host variable and verbose cleanup

c465ccd remove ineffassign check

398f764 Merge pull request #104 from jun06t/fluent

8dffe6f Merge pull request #103 from scaleoutsean/patch-1

59c98e1 Merge pull request #109 from dupondje/updates

403872d Fix metricId's when using multiple vCenter (versions)

e9a2141 Fix some typos

9711115 Allow configuration of ResultLimit and InstanceRatio

9966873 Enabled at-least-once message transferring

cb44c56 Trivial typo

aba709b (tag: v0.8a) Merge pull request #95 from cblomart/goroutines
```

In what section of the config file do the 2 extra options belong ?

**Comment by cblomart** [comment]
The parameters would need to be set in the root of the configuration file.
I made a new push to prometheus_race to set defaults for these configurations (commit 259d562). So they are not mandatory and don't break backward compatibility.

**Comment by dupondje** [comment]
https://github.com/cblomart/vsphere-graphite/commit/d1d2c85cef55d6b90eb6be79ab8035a26a85467d

This should make things easier :)
Still needs testing, but looks fine with first tests!

**Comment by cblomart** [comment]
Ok, i've merged the changes of @dupondje  in the prometheus_race branch. There is also a docker container image available "cblomart/vsphere-graphite:875a98b".

I would like to wait for some test results before merging this in the main branch.

**Comment by DirkFries** [comment]
Hi Cedric,

after putting the parameters in the config file my 2 vsphere-graphite processes run fine and metrics are beeing gathered.
👍 

For the records: My highest mark of go routines was 625 in march.

**Comment by cblomart** [comment]
Hi Dirk, 

Good to know it does work now.

The changes from @dupondje did go further than what was initially in the branch so be sure to test with the latest version (of the branch).

Many goroutines is a bad thing at first sight. I think the main goal is for them not to grow!
This growth is a reason for goroutine leak leading to a memory leak.

So: can you test with the latest version in there, ideally for a few days and look if memory is ok as well as goroutines (should be steady)?

Cédric

**Comment by dupondje** [comment]
For me it did not only fix some occasional blocks, but also mem leaks.

**Comment by DirkFries** [comment]
Hi Cedric,

sorry, but I got some problems compiling.
This is my approach:

```
cd /tmp
git clone https://github.com/cblomart/vsphere-graphite.git
cd vsphere-graphite
git checkout prometheus_race
export GOROOT=~/golang
export GOPATH=~/go
export PATH=~/golang/bin:$PATH
make
```

But "make" fails with

```
host:/tmp/vsphere-graphite # make
go get -u honnef.co/go/tools/cmd/staticcheck
package flag: unrecognized import path "flag" (import path does not begin with hostname)
package fmt: unrecognized import path "fmt" (import path does not begin with hostname)
package go/ast: unrecognized import path "go/ast" (import path does not begin with hostname)
package go/token: unrecognized import path "go/token" (import path does not begin with hostname)
package go/types: unrecognized import path "go/types" (import path does not begin with hostname)
package bytes: unrecognized import path "bytes" (import path does not begin with hostname)
package reflect: unrecognized import path "reflect" (import path does not begin with hostname)
package unicode: unrecognized import path "unicode" (import path does not begin with hostname)
package encoding/gob: unrecognized import path "encoding/gob" (import path does not begin with hostname)
package encoding/hex: unrecognized import path "encoding/hex" (import path does not begin with hostname)
```

**Comment by cblomart** [comment]
Some golang related issue i suppose.
Which version of golang do you have ?

**Comment by DirkFries** [comment]
Hi Cedric,

I tried on 2 machines

SLES 15 SP1
go version go1.12.17 linux/amd64

Ubuntu 19.10
go version go1.12.10 linux/amd64

**Comment by cblomart** [comment]
You are trying to build out of your gopath.

Try:
```bash
> mkdir -p $GOPATH/src/github.com/cblomart/
> cd $GOPATH/src/github.com/cblomart/
> git clone https://github.com/cblomart/vsphere-graphite.git
> make
```

**Comment by cblomart** [comment]
Solution may be go modules but the project is not there yet...

https://blog.golang.org/using-go-modules

As of Go 1.11, the go command enables the use of modules when the current directory or any parent directory has a go.mod, provided the directory is outside $GOPATH/src. (**Inside $GOPATH/src, for compatibility, the go command still runs in the old GOPATH mode, even if a go.mod is found.** See the go command documentation for details.) Starting in Go 1.13, module mode will be the default for all development.

**Comment by cblomart** [comment]
I have just added go module to the last commit to the branch.
you should be able to build with the make now.

**Comment by DirkFries** [comment]
OK, got it finally build 
Thanks a lot 👍 

Well looks like it is working fine, Metrics are beeing gathered and I have about 15 go routines.

Shall I report again in about a week ?

**Comment by cblomart** [comment]
Looks good! Maybe you can check in a few days already.

**Comment by DirkFries** [comment]
Hi Cedric,

since changing the binary vsphere-graphite is running fine. go routines of one process have risen to 16, the other one is still at 15.

I will report back on monday

**Comment by cblomart** [comment]
Looks good 👍 
Awaiting Monday then!

**Comment by DirkFries** [comment]
vsphere-graphite is still working fine.  Routines of one process is still at 15, routines of the other one have risen to 17.

**Comment by cblomart** [comment]
Good LGTM so far.
I will merge the branch.

**Comment by cblomart** [comment]
Thanks to @DirkFries and @dupondje this should be solved now!

**Comment by DirkFries** [comment]
Hi @all,

unfortunaly it look like it happened again today.
vsphere-graphite had about 50% memory

Tasks: 271 total,   1 running, 270 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0,0 us,  0,0 sy,  0,0 ni,100,0 id,  0,0 wa,  0,0 hi,  0,0 si,  0,0 st
KiB Mem:  16240800 total, 16051460 used,   189340 free,   709708 buffers
KiB Swap:  2103292 total,        0 used,  2103292 free.  4680060 cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
10313 prometh+  20   0 8966152 7,638g   8072 S 0,000 49,31   1849:34 /data/vsphere-graphite/vsphere-graphite

and a lot of Go Routines:

2020-05-13T09:37:10.957851+02:00 lin372 vsphere-graphite[10313]: 2020/05/13 09:37:10 go routines: 271
2020-05-13T09:37:10.958099+02:00 lin372 vsphere-graphite[10313]: 2020/05/13 09:37:10 go routines: 270
2020-05-13T09:37:10.958350+02:00 lin372 vsphere-graphite[10313]: 2020/05/13 09:37:10 go routines: 269
2020-05-13T09:37:10.958617+02:00 lin372 vsphere-graphite[10313]: 2020/05/13 09:37:10 go routines: 268
2020-05-13T09:37:10.958869+02:00 lin372 vsphere-graphite[10313]: 2020/05/13 09:37:10 go routines: 267
2020-05-13T09:37:10.959125+02:00 lin372 vsphere-graphite[10313]: 2020/05/13 09:37:10 go routines: 266
2020-05-13T09:37:10.959401+02:00 lin372 vsphere-graphite[10313]: 2020/05/13 09:37:10 go routines: 265
2020-05-13T09:37:10.959666+02:00 lin372 vsphere-graphite[10313]: 2020/05/13 09:37:10 go routines: 264
2020-05-13T09:37:10.959947+02:00 lin372 vsphere-graphite[10313]: 2020/05/13 09:37:10 go routines: 263

Unfortunaly I forgot to check of the Webserver times out while collecting metrics with Prometheus. So I can not guarantee that, but I did now get any metrics in Prometheus, so I think it's highly probable, that Prometheus got a timeout.

Bye Dirk

**Comment by cblomart** [comment]
@DirkFries 

Can you confirm the version you are using?

concerning the go routines what perturbs me is that the same daemon logs multiple times the goroutines at the same time... would there be multiple metrics requests at the same time? (here 2020/05/13 09:37)

Please don't try to scrape vsphere-grahite more than once a minute, there is no point doing so.

**Comment by DirkFries** [comment]
Hi Cedric,

the version I use is the one UI build on 31 Mar. with your support.
Unfortunaly it shows no version

host:/home/vsphere # ./vsphere-graphite --version
2020/05/13 10:56:57 Version information: NA - NA@NA (NA)
Usage: vsphere-graphite install | remove | start | stop | status

The mutiple connects are an interesting point. In my Prometheus Config Vsphere-Graphite is checked every 60s.
And under normal circumstances the logfiles confirms that:

2020-05-13T10:52:04.969827+02:00 host vsphere-graphite[5571]: 2020/05/13 10:52:04 go routines: 26
2020-05-13T10:53:04.959156+02:00 host vsphere-graphite[5571]: 2020/05/13 10:53:04 go routines: 26
2020-05-13T10:54:04.929824+02:00 host vsphere-graphite[5571]: 2020/05/13 10:54:04 go routines: 26

Bye, Dirk

**Comment by cblomart** [comment]
Ok, to be sure of the version, i will make a release and you can eventually use the binary from it (will update this thread with the URL to get it).
-> not saying that it will change anything tough

**Comment by cblomart** [comment]
ok... the updated binary: https://github.com/cblomart/vsphere-graphite/releases/download/v0.8b/vsphere-graphite_linux_amd64.tgz

**Comment by DirkFries** [comment]
Hi Cedric,

thank you a lot. 
I just downloaded the binary, installed and restartet and I'll report in the future, although it might take some months.

host:/home/vsphere # ./vsphere-graphite --version
2020/05/18 10:30:26 Version information: NA - 37a9dcb@HEAD (dirty)
Usage: vsphere-graphite install | remove | start | stop | status

To see how the memory consumption raises I have prepared a Graph. So maybe next time we can see if memory consumption raises slowly over time or suddenly.

Bye, Dirk

**Comment by dupondje** [comment]
It would also be nice to have a list of what the goroutines are doing when they use that much memory.
A mem pprof could help for example.

**Comment by cblomart** [comment]
I am not sure it is a best practice but there is a configuration you can set in the config to have memory profile: "MEMProfiling: true"
It will on each collection dump the mem profile to "/tmp/vsphere-graphite-mem.pb.gz"

**Comment by DirkFries** [comment]
Hi @all,

I have activated MEMProfiling: true and I am saving a sample once an hour because I am pretty sure I will forget to save the last file before a restart if the problem should occur again in some months ;-)

Bye, Dirk

**Comment by DirkFries** [comment]
Hi @all,

vsphere-graphite went wild again today:

2020-05-19T11:07:35.936472+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:07:35 go routines: 576
2020-05-19T11:08:35.931149+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:08:35 go routines: 75
2020-05-19T11:09:35.907987+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:09:35 go routines: 75
2020-05-19T11:10:36.025229+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:10:36 go routines: 75
2020-05-19T11:11:35.930051+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:11:35 go routines: 75
2020-05-19T11:12:35.920192+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:12:35 go routines: 75
2020-05-19T11:13:36.052244+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:13:36 go routines: 75
2020-05-19T11:14:35.993959+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:14:35 go routines: 75
2020-05-19T11:15:35.969436+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:15:35 go routines: 78
2020-05-19T11:16:36.046579+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:16:36 go routines: 75
2020-05-19T11:17:36.070426+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:17:36 go routines: 75
2020-05-19T11:18:35.924376+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:18:35 go routines: 110
2020-05-19T11:19:35.965830+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:19:35 go routines: 917
2020-05-19T11:19:36.434226+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:19:36 go routines: 70
2020-05-19T11:20:36.027085+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:20:36 go routines: 1175
2020-05-19T11:20:36.475763+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:20:36 go routines: 64
2020-05-19T11:21:36.072054+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:21:36 go routines: 851
2020-05-19T11:21:36.468112+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:21:36 go routines: 58
2020-05-19T11:22:36.161588+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:22:36 go routines: 891
2020-05-19T11:22:36.446311+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:22:36 go routines: 52
2020-05-19T11:23:36.154339+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:23:36 go routines: 1007
2020-05-19T11:23:36.539769+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:23:36 go routines: 46
2020-05-19T11:24:36.034874+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:24:36 go routines: 1138
2020-05-19T11:24:36.470442+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:24:36 go routines: 40
2020-05-19T11:25:36.076535+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:25:36 go routines: 807
2020-05-19T11:25:36.284589+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:25:36 go routines: 34
2020-05-19T11:26:35.999414+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:26:35 go routines: 991
2020-05-19T11:26:36.517554+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:26:36 go routines: 28
2020-05-19T11:27:36.053788+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:27:36 go routines: 861
2020-05-19T11:27:36.409133+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:27:36 go routines: 22
2020-05-19T11:28:36.119689+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:28:36 go routines: 832
2020-05-19T11:28:36.414930+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:28:36 go routines: 16

Unfortunaly I collect a sample MEMProfiling File every hour at 5 Minutes past the full hour and at this time the show over again.

RSS Usage hasen't changed much:

![grafik](https://user-images.githubusercontent.com/51441256/82316905-ccb54a80-99cd-11ea-9866-85e720d90af0.png)

-> I will collect the MEMProfiling File now every minute

**Comment by dupondje** [comment]
I dont see the issue?
Its normal you have a high goroutine count when multiple collections are running at the same time.
The moment the amount of goroutines is printed, and another collection is ongoing, then you will notice this high number.

The good thing is that it drops, so collection does not hang somewhere. So this looks perfectly fine!

**Comment by cblomart** [comment]
@dupondje is true there... looks that goroutines go away so there is no "leak" per se.

Anyways what I don't get is why are there multiple queries in there... are you collecting from different Prometheus(es)? 

As today there is no way to filter Prometheus query output, the only relevant reason I see to request multiple times at the same time is to send to different backends...

**Comment by DirkFries** [comment]
I am only querying from one Prometheus, every 60 seconds.

But in the vsphere-graphite config a have 3 vspheres with about 1200 hosts.
Maybe I should split my vsphere-graphite config in 3 single processes ?

**Comment by DirkFries** [comment]
Ok, instead of 2 vpshere-graphite processes I have now split it into 4 processes. 
To avoid confusion: I had already split my config into 2 processes because of different vcenter versions ( https://github.com/cblomart/vsphere-graphite/issues/114 )

@dupondje 

I rechecked my logs and you can see, that Prometheus scrapes nicely every 60s.
But between 10:54 and 10:59 and 10:59 and 11:07 there are missing checks.
I can only guess, that one of the 3 vcenters is laggy and keeps vsphere-graphite waiting at that point and that this is the reason for all this trouble.
Maybe a timeout option would help. I am not sure if there is one, because I found none in vsphere-graphite-example.json

2020-05-19T10:39:36.564777+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:39:36 go routines: 15
2020-05-19T10:40:36.337976+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:40:36 go routines: 15
2020-05-19T10:41:36.248596+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:41:36 go routines: 15
2020-05-19T10:42:36.245336+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:42:36 go routines: 15
2020-05-19T10:43:36.316304+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:43:36 go routines: 15
2020-05-19T10:44:36.629189+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:44:36 go routines: 15
2020-05-19T10:45:36.207135+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:45:36 go routines: 15
2020-05-19T10:46:36.362862+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:46:36 go routines: 15
2020-05-19T10:47:36.248827+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:47:36 go routines: 15
2020-05-19T10:48:36.324050+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:48:36 go routines: 15
2020-05-19T10:49:36.523156+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:49:36 go routines: 15
2020-05-19T10:50:36.466820+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:50:36 go routines: 15
2020-05-19T10:51:36.305647+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:51:36 go routines: 15
2020-05-19T10:52:36.323104+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:52:36 go routines: 15
2020-05-19T10:53:36.115900+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:53:36 go routines: 15
2020-05-19T10:54:36.282624+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:54:36 go routines: 15
2020-05-19T10:59:17.473059+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:59:17 go routines: 33
2020-05-19T10:59:21.454027+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 10:59:21 go routines: 28
2020-05-19T11:07:35.936472+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:07:35 go routines: 576
2020-05-19T11:08:35.931149+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:08:35 go routines: 75
2020-05-19T11:09:35.907987+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:09:35 go routines: 75
2020-05-19T11:10:36.025229+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:10:36 go routines: 75
2020-05-19T11:11:35.930051+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:11:35 go routines: 75
2020-05-19T11:12:35.920192+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:12:35 go routines: 75
2020-05-19T11:13:36.052244+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:13:36 go routines: 75
2020-05-19T11:14:35.993959+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:14:35 go routines: 75
2020-05-19T11:15:35.969436+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:15:35 go routines: 78
2020-05-19T11:16:36.046579+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:16:36 go routines: 75
2020-05-19T11:17:36.070426+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:17:36 go routines: 75
2020-05-19T11:18:35.924376+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:18:35 go routines: 110
2020-05-19T11:19:35.965830+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:19:35 go routines: 917
2020-05-19T11:19:36.434226+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:19:36 go routines: 70
2020-05-19T11:20:36.027085+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:20:36 go routines: 1175
2020-05-19T11:20:36.475763+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:20:36 go routines: 64
2020-05-19T11:21:36.072054+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:21:36 go routines: 851
2020-05-19T11:21:36.468112+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:21:36 go routines: 58
2020-05-19T11:22:36.161588+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:22:36 go routines: 891
2020-05-19T11:22:36.446311+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:22:36 go routines: 52
2020-05-19T11:23:36.154339+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:23:36 go routines: 1007
2020-05-19T11:23:36.539769+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:23:36 go routines: 46
2020-05-19T11:24:36.034874+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:24:36 go routines: 1138
2020-05-19T11:24:36.470442+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:24:36 go routines: 40
2020-05-19T11:25:36.076535+02:00 lin372 vsphere-graphite[32664]: 2020/05/19 11:25:36 go routines: 807

Let's see how it works in the future ...

**Comment by cblomart** [comment]
Hello, I hope you are all well and safe!

On the front of goroutines, do you still see go high go routines values @DirkFries ?

vsphere-graphite uses a simple context with cancelation today:
```golang
ctx, cancel := context.WithCancel(context.Background())
```

I am wondering if I should replace that with a deadline context (1 minute) to timeout after some time.
https://golang.org/src/context/context.go?s=13290:13358#L417

I won't be able to test directly so if someone is willing to test I am open to create a branch with this modification.

Cédric

PS: and thanks for the reminder on the different sphere versions issue!

**Comment by cblomart** [comment]
@DirkFries 
Where are you on the test with different instances? is it more stable?
Would you be interested in testing/validating the context change?

Cédric

**Comment by DirkFries** [comment]
Hi Cedric,

sorry for the delay.
Right now everything works pretty fine.

In fact my installations are in productive environment, so testing is always a little difficult.
But I think I can query one of the vcenters during certain time periods with a binary with deadline context and slowly increase testing time and eventually use it for other vcenters, too.

To put a long story short : Yes :-)

**Comment by cblomart** [comment]
@DirkFries I will close for now. If any more concerns you know you can react on the issue :)
